docs/code/CrossPlatform.md

```diff
@@ -0,0 +1,175 @@
+
+ML.NET also has dependencies on things that either don't build on other architectures or have to be compiled by the user if it's wanted. For example:
+ - LightGBM
+ - TensorFlow
+ - ONNX Runtime
+
+I will go over these in more depth below.
+
+### 2.2 Build
+Since ML.NET has a hard dependency on x86/x64, the build process assumes it's running there. For example, the build process will try and copy native DLLs without checking if they exist because it assumes the build for them succeeded or that they are available. The build process will need to be modified so that it doesn't fail when it can't find these files. It does the same copy for our own Native DLLs, so this will need to be fixed for those as well.
+
+### 2.3 Managed Code
+Since ML.NET has a hard dependency on x86/x64, the managed code imports DLLs without checking whether or not they exist. If the DLLs don't exist you get a hard failure. For example, if certain columns are active, the `MulticlassClassificationScorer` will call `CalculateIntermediateVariablesNative` which is loaded from `CpuMathNative`, but all of this is done without any checks to see if the DLL actually exists. The tests also run into this problem, for instance, the base test class imports and sets up Intel MKL even if the test itself does not need it.
+
+### 2.4 Native Projects
+ML.NET has 6 native projects. They are:
+ - CpuMathNative
+   - Partial managed fallback when using NetCore 3.1.
+   - A large amount of work would be required to port the native code to other platforms. We would have to change all the SIMD instructions for each platform.
+   - A small amount of work required for a full managed fallback.
+   - This was created before we had hardware intrinsics so we used native code for performance. The managed fallback uses x86/x64 intrinsics if the hardware supports it, otherwise it has a plain managed fallback without intrinsics if needed.
+ - FastTreeNative
+   - Full managed fallback by changing a C# compiler flag. This flag is hardcoded to always use the native code.
+   - Small amount of work required to change build process and verify it's correct.
+   - This was created before we had hardware intrinsics so we used native code for performance. The managed fallback does not currently use hardware intrinsics, so this will be slower than the native solution. Hardware intrinsics could be added to improve performance.
+ - LdaNative
+   - No managed fallback, but builds successfully on non x86/x64 without changes.
+   - Large amount of work to have a managed fallback. This solution is about 3000 lines of code not including any code from the dependencies.
```

Comment:
Question for this and the other libraries where we might have to write a bunch of managed code -- how good are the tests? If there are poor tests, that will add risk/cost because they will have to be improved before major work.

Reply:
So the native code has no testing itself, but the managed code that calls the native code has good/decent test coverage. Enough that if we do something wrong during this process it should be caught. Or at least enough to catch it in the majority of the cases. Overall we have about 70% test coverage, so it should be enough to catch the majority of them.